### PR TITLE
driver/clock_control: Clean up stm32h7 and stm32u5 drivers

### PR DESCRIPTION
**Pull request by erwango**
Attempt to clean up STM32 clock_control driver, starting on H7 and U5 series.
Main target is to clean up `clock_stm32_ll_common.c` which is used by all the other STM32 series,
but I first want to validate ideas on these 2 first ones.

In the end, there are not much changes, aside from:
- factorization when possible
- use of IS_ENABLED when possible to ease readability

Changes were tested thanks to new dedicated tests available in `tests/drivers/clock_control/stm32_clock_configuration`.

Based on:
- ~~https://github.com/zephyrproject-rtos/zephyr/pull/41310~~ (merged)
- ~~https://github.com/zephyrproject-rtos/zephyr/pull/41391~~ (merged)

EDIT (01/18):
Following changes were made on top of existing proposal. Reason not to split this in another PR is that I prefer to group breaking changes in one big PR and then be done with it. I think this is also preferable for driver users.
- Reworked various clocks sources configuration and enabling. They're now configured independently and in logical order (fixed clocks, then PLLs) at startup based on device tree status. Then they are used if required when configuring sysclock
- Reworked some STM32_CLOCK dt symbols in order to simplify usage in clock_control drivers.

The rationale behind this new changes is to prepare introduction of alternate / optional clock sources (cf #41650), which requires to be able to enable each source clock independently from each others.

### CODE REVIEW

**Review by ABOSTM — CHANGES_REQUESTED** [review]
Few inline comments, otherwise, LGTM

**Comment by erwango** [comment]
@ABOSTM Thanks for the review. Comments updated.

**Review by ABOSTM — DISMISSED** [review]
LGTM

File: drivers/clock_control/clock_stm32_ll_u5.c [hunk 1]
```diff
@@ -51,29 +51,29 @@ void config_enable_default_clocks(void)
...
+		while (!LL_RCC_LSE_IsReady()) {
+		}
 
-	/* Enable LSESYS additionnally */
-	SET_BIT(RCC->BDCR, RCC_BDCR_LSESYSEN);
-	/* Wait till LSESYS is ready */
-	while (READ_BIT(RCC->BDCR, RCC_BDCR_LSESYSRDY) == 0U) {
-	}
+		/* Enable LSESYS additionnally */
+		SET_BIT(RCC->BDCR, RCC_BDCR_LSESYSEN);
```
Comment (FRASTM) [inline drivers/clock_control/clock_stm32_ll_u5.c:70]: this is `LL_RCC_LSE_EnablePropagation()` in stm32U5

File: drivers/clock_control/clock_stm32_ll_u5.c [hunk 2]
```diff
@@ -51,29 +51,29 @@ void config_enable_default_clocks(void)
...
+		/* Wait till LSESYS is ready */
+		while (READ_BIT(RCC->BDCR, RCC_BDCR_LSESYSRDY) == 0U) {
```
Comment (FRASTM) [inline drivers/clock_control/clock_stm32_ll_u5.c:72]: this is `LL_RCC_LSESYS_IsReady()` in stm32U5

File: drivers/clock_control/clock_stm32_ll_h7.c [hunk 3]
```diff
@@ -510,102 +506,97 @@ static struct clock_control_driver_api stm32_clock_control_api = {
...
-	/* Main PLL configuration and activation */
-	LL_RCC_PLL_SetSource(LL_RCC_PLLSOURCE_CSI);
-
-#elif STM32_PLL_SRC_HSI
-	/* By default choose HSI as PLL clock source */
-
-	/* Enable HSI oscillator */
-	LL_RCC_HSI_Enable();
-	while (LL_RCC_HSI_IsReady() != 1) {
+	LL_RCC_PLL3_Enable();
```
Comment (FRASTM) [inline drivers/clock_control/clock_stm32_ll_h7.c:548]: could we have 	`if (IS_ENABLED(STM32_HSI_ENABLE)) {`

File: drivers/clock_control/clock_stm32_ll_h7.c [hunk 4]
```diff
@@ -510,102 +506,97 @@ static struct clock_control_driver_api stm32_clock_control_api = {
...
-#endif /* ! CONFIG_CPU_CORTEX_M4 */
-
-	ARG_UNUSED(dev);
 
-#if !defined(CONFIG_CPU_CORTEX_M4)
+	/* Initialize PLL 3 */
+	r = get_vco_input_range(STM32_PLL3_M_DIVISOR, &vco_input_range);
+	if (r < 0) {
+		return r;
+	}
```
Comment (FRASTM) [inline drivers/clock_control/clock_stm32_ll_h7.c:521]: Is the `stm32_clock_control_init` function returning 0 ?
In the clock_stm32_ll_common.c we read:
```
/**
 * @brief Initialize clocks for the stm32
 *
 * This routine is called to enable and configure the clocks and PLL
 * of the soc on the board. It depends on the board definition.
 * This function is called on the startup and also to restore the config
 * when exiting for low power mode.
 *
 * @param dev clock device struct
 *
 * @return 0
 */
```

**Review by FRASTM — COMMENTED** [review]
The stm32_clock_control_init function return code seems not tested yet, however if it could be <>0 , then we could see in the function something like 
```
if (IS_ENABLED(STM32_PLL_SRC_HSE)) {
...
if (IS_ENABLED(STM32_PLL3_P_ENABLE)) {
...
else
return -ENOTSUP
```

File: drivers/clock_control/clock_stm32_ll_h7.c (hunk 4 above)
Comment (erwango) [reply drivers/clock_control/clock_stm32_ll_h7.c:521]: We'll probably need to fix this description, but I'll address this when reviewing this driver.

**Review by FRASTM — DISMISSED** [review]

**Comment by erwango** [comment]
DNM as this will likely require more work in short term to support multiple/alternate peripheral source clocks.

File: drivers/clock_control/clock_stm32_ll_h7.c [hunk 5]
```diff
@@ -510,102 +506,117 @@ static struct clock_control_driver_api stm32_clock_control_api = {
...
+	LL_RCC_PLL3_Enable();
+	while (LL_RCC_PLL3_IsReady() != 1U) {
+	}
+#endif /* STM32_PLL3_ENABLE */
 
-#if STM32_HSE_BYPASS
-	LL_RCC_HSE_EnableBypass();
-#else
-	LL_RCC_HSE_DisableBypass();
-#endif
```
Comment (erwango) [inline drivers/clock_control/clock_stm32_ll_h7.c:565]: This is now missing

Comment (erwango) [reply drivers/clock_control/clock_stm32_ll_h7.c:574]: Has now been fixed

File: drivers/clock_control/clock_stm32_ll_h7.c [hunk 6]
```diff
@@ -501,163 +501,163 @@ static struct clock_control_driver_api stm32_clock_control_api = {
...
-		/* Main PLL configuration and activation */
-		LL_RCC_PLL_SetSource(LL_RCC_PLLSOURCE_CSI);
-	} else if (IS_ENABLED(STM32_PLL_SRC_HSI)) {
-		set_up_hsi();
-
-		/* Main PLL configuration and activation */
-		LL_RCC_PLL_SetSource(LL_RCC_PLLSOURCE_HSI);
-	} else {
-		return -ENOTSUP;
+	if (IS_ENABLED(STM32_PLL_ENABLED) || IS_ENABLED(STM32_PLL3_ENABLED)) {
```
Comment (ABOSTM) [inline drivers/clock_control/clock_stm32_ll_h7.c:524]: `if (IS_ENABLED(STM32_PLL_ENABLED) || IS_ENABLED(STM32_PLL3_ENABLED)) {`
Is redundant with 
`#if defined(STM32_PLL_ENABLED) || defined(STM32_PLL3_ENABLED)`
few lines above

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: drivers/clock_control/clock_stm32_ll_h7.c (hunk 6 above)
Comment (erwango) [reply drivers/clock_control/clock_stm32_ll_h7.c:524]: Good catch, removed, thanks

**Review by ABOSTM — DISMISSED** [review]

**Review by FRASTM — DISMISSED** [review]
change commit msg to drivers/clock_control: stm32u5: Use LL API for L**S**ESYS programming

**Comment by erwango** [comment]
> change commit msg to drivers/clock_control: stm32u5: Use LL API for L**S**ESYS programming

Fixed, Thanks

**Review by FRASTM — APPROVED** [review]

**Review by ABOSTM — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**